src/v2/Apps/Conversation/Components/ConversationMessages.tsx

```diff
@@ -1,60 +1,133 @@
-import { groupMessages } from "../Utils/groupMessages"
+import { groupMessages, Message as MessageType } from "../Utils/groupMessages"
 import { TimeSince, fromToday } from "./TimeSince"
 import { MessageFragmentContainer as Message } from "./Message"
-import React from "react"
+import React, { useEffect, useState } from "react"
 import { Spacer } from "@artsy/palette"
 import { graphql } from "relay-runtime"
 import { createFragmentContainer } from "react-relay"
 import { ConversationMessages_messages } from "v2/__generated__/ConversationMessages_messages.graphql"
-import compact from "lodash/compact"
+import { ConversationMessages_events } from "v2/__generated__/ConversationMessages_events.graphql"
+
+import { extractNodes } from "v2/Utils/extractNodes"
+import { DateTime } from "luxon"
+import { sortBy } from "lodash"
+import { OrderUpdateFragmentContainer } from "./OrderUpdate"
+import { OrderUpdate_event } from "v2/__generated__/OrderUpdate_event.graphql"
+import { Message_message } from "v2/__generated__/Message_message.graphql"
+
 
 interface ConversationMessageProps {
   messages: ConversationMessages_messages
+  events: ConversationMessages_events | null
 }
+type Order = NonNullable<
+  NonNullable<
+    NonNullable<NonNullable<ConversationMessages_events>["edges"]>[number]
+  >["node"]
+>
+type OrderEvent = Order["orderHistory"][number]
 
 export const ConversationMessages = ({
   messages,
+  events,
 }: ConversationMessageProps) => {
+  let [messagesAndEvents, setMessagesAndEvents] = useState<MessageType[][]>([])
+
+  useEffect(() => {
+    const allMessages = extractNodes(messages)
+    const allOrderEvents = extractNodes(events).reduce<OrderEvent[]>(
+      (prev, order) => prev.concat(order.orderHistory),
+      []
+    )
+    const orderEventsWithoutFailedPayment = allOrderEvents.filter(
+      (event, index) => {
+        if (
+          !!event &&
+          !(
+            event.state === "APPROVED" &&
+            allOrderEvents[index + 1] &&
+            allOrderEvents[index + 1].state === "SUBMITTED"
+          )
+        ) {
+          return event
+        }
+      }
+    )
+    const sortedMessages = sortBy(
+      [...orderEventsWithoutFailedPayment, ...allMessages],
+      message => DateTime.fromISO(message.createdAt!)
+    )
+    const groupAllMessages = groupMessages(sortedMessages)
+    setMessagesAndEvents(groupAllMessages.reverse())
+  }, [messages, events])
+
+  const relevantEvents = [
+    "CommerceOfferSubmittedEvent",
+    "CommerceOrderStateChangedEvent",
+  ]
+  type DisplayableMessage = OrderUpdate_event | Message_message
```

Comment:
minor: the `type` here mixed in with code is a little awkward to me. Some may disagree with me though if it is very specific to this section. Non-blocking!

Reply:
Thank you for this note! Make a sense!